### ISSUE TITLE
Decouple dataset class from algorithms and models - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#212]
_Merges decouple-dataset-class -> dev_

### What does the code in the MR do ?

This MR proposes to start the work on decoupling the different modules of Leaspy. It starts with a relatively easy one: decoupling the `Dataset` class from `Algorithm` and `Model` classes.

Previous to this MR, the `Dataset` constructor could receive an instance of a model and algorithm:

```python
dataset = Dataset(data, algo=algorithm, model=model)
```

Unless I missed something, the only reason the `Dataset` class needs to know about models and algorithms is to run some compatibility checks: `_check_model_compatibility` and `_check_algo_compatibility`. The later does basically nothing, and the former seems more like a `Model`'s responsibility than a `Dataset`'s one. These checks are already partially implemented in the `initialize` method of the `Model` class.

This MR proposes to remove these checks from the `Dataset` class such that the constructor can be simplified in:

```python
dataset = Dataset(data)
```

### When is the MR due for? (review deadline)

No deadline.

### What issues are linked to the MR ?

No linked issue.

**Comment by NicolasGensollen** [comment]
added 103 commits

<ul><li>72b758ef...2e6e8fe9 - 100 commits from branch <code>dev</code></li><li>b966e892 - remove the algo and model checking by the dataset class</li><li>44a974ff - Type hints and PEP8 stuffs</li><li>d04c2cef - update tests</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=596167385&start_sha=72b758ef3139d2f0b445751a936899e2262ee892)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>8ead8bb3 - implement validate_compatibility_of_dataset for all models</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=596401244&start_sha=1eac7770fcf4b887ed380107c788ea6771f9ac70)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>7703f065 - fix test before merging univariate refacto PR</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=598049710&start_sha=8ead8bb3d8c7c3ab55db1fc66138252b5c1bf070)

**Comment by NicolasGensollen** [comment]
added 8 commits

<ul><li>7703f065...98e535ea - 2 commits from branch <code>dev</code></li><li>c994e75d - remove the algo and model checking by the dataset class</li><li>1d85931e - Type hints and PEP8 stuffs</li><li>6f8a2ffa - update tests</li><li>704721bb - remove conditional initialization through api fit method</li><li>71047e4f - implement validate_compatibility_of_dataset for all models</li><li>d76d3acd - fix test before merging univariate refacto PR</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604282080&start_sha=7703f0654a70ae81f895ecd9287dc62c3e497c73)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>160c2670 - remove unecessary check</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604350638&start_sha=d76d3acd8aceac1a99a3307d910aba4f0c17c1c5)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/algo/simulate/simulate.py line 791](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..589de967af104f2935cd3381e5062a371a74e677#diff-2ff2f8603288df9a1d718aa3a2305db4R791)



1. Je ne comprends pas l'ancien commentaire qui traîne au-dessus de ta modif mais j'ai l'impression que tu peux le virer.
2. You should add the new `model.validate_....(dataset)` after creation of the dataset no?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/io/data/dataset.py line 12](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..160c2670f73d5227679c1da86116dd434219f82b#diff-38ba270a205908168d6cae65b7af44feR12)



Importé juste au-dessus non?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/api.py line 142](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..80c96ee228435966dc5cbc4d5a8f029fcc49d783#diff-02af7144d340c497a7e9c91880133566L142)



This comment is still legit so do not remove it please (or move it elsewhere if you want but not really intended to be in public docs - mostly for future dev).

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/models/abstract_multivariate_model.py line 81](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..589de967af104f2935cd3381e5062a371a74e677#diff-aa3a0fa1e9d2af1db4aea3e12a4e9d5dR81)



I think you can remove `and self.features is not None` from the condition here (it was added for the very special `ConstantModel` - cf. comment in `GenericModel`)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/models/abstract_multivariate_model.py line 102](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..bca3e52ec6ab6fdaa9b0b6f2fdd113a2da7b7075#diff-aa3a0fa1e9d2af1db4aea3e12a4e9d5dR102)



Actually those conditions should be checked for all models (including special models Constant & LME)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/api.py line 144](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..1d0dbba2d5bc0f7be5f201421d4f42cbb821adbb#diff-02af7144d340c497a7e9c91880133566R144)



Actually I'm wondering whether when should not `validate_compatibility_of_dataset` before initialization of model to fail fast in cases where they are not compatible (e.g. univariate model or LMEModel with a dataset with n>1 features)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:08

Commented on [leaspy/models/generic_model.py line 98](https://github.com/aramis-lab/leaspy/compare/98e535eaa9f05aeba4af003a6d0a84653d7dc8eb..589de967af104f2935cd3381e5062a371a74e677#diff-060e865e3cdeecf0ff6d8b322e8e9d4eR98)



Everything is a bit circular / dirty here, because in GenericModel, model.dimension is a property = length of model.features (or None if features is None), but then for LME Model we could have `dimension=2` and `features=['A','B'] at the end of this routine which is not very satisfactory.

- May be we should hardcode: read-only (class) property dimension=1 for LMEModel (and for manifold univariate model?)
- Actually it may be cleaner to set `self.features = None` when we say that we are overwriting previous features, then do the compat check and then store new features only after compat check pass (so not to leave strange `features` set in model)
- Same remarks apply in abstract multivariate model in fact

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 15, 2023, 14:12

Final note: currently `AbstractAlgo.run` has a strange signature (because it tries to cover all 3 different types of algos: fit, personalization, simulation), but for the future, since all of them take a model + dataset as input I guess we could add the `model.validate_compat...(dataset)` inside the abstract algo run method?

**Comment by NicolasGensollen** [comment]
changed this line in [version 8 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604718867&start_sha=160c2670f73d5227679c1da86116dd434219f82b#e90f12a2006dc68894a8bd4a69b3827b0a3ab992_12_12)

**Comment by NicolasGensollen** [comment]
added 2 commits

<ul><li>8631fefc - remove useless import</li><li>ff9bd87c - remove None features check</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604718867&start_sha=160c2670f73d5227679c1da86116dd434219f82b)

**Comment by NicolasGensollen** [comment]
Indeed, thanks

**Comment by NicolasGensollen** [comment]
Remove it in ff9bd87c

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>80c96ee2 - remove cryptic comment and validate the compatibility of the dataset</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604735749&start_sha=ff9bd87ca0c919dea270bc3992d6a404fc79777f)

**Comment by NicolasGensollen** [comment]
I removed the comment and added the validity check in 80c96ee2

Thanks for catching this.

**Comment by NicolasGensollen** [comment]
changed this line in [version 10 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604741361&start_sha=80c96ee228435966dc5cbc4d5a8f029fcc49d783#15761437d39369b398641b30e6eba7ec0289a59c_140_142)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>bca3e52e - add comment back</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=604741361&start_sha=80c96ee228435966dc5cbc4d5a8f029fcc49d783)

**Comment by NicolasGensollen** [comment]
Fair enough ! I added it back in bca3e52e

**Comment by NicolasGensollen** [comment]
changed this line in [version 11 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605524551&start_sha=bca3e52ec6ab6fdaa9b0b6f2fdd113a2da7b7075#b8d605d737e5d8c90331a65ddbb63c47476f8c90_102_79)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>c38e5099 - move validate_dataset_compat to parent classes and make dimension and features attributes private</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605524551&start_sha=bca3e52ec6ab6fdaa9b0b6f2fdd113a2da7b7075)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>beca27ae - Define BaseModel and move generic logic inside it</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605571440&start_sha=c38e509916c5b843c291b283c4c92bc02ad9acca)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>1d0dbba2 - Define BaseModel and move generic logic inside it</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605586678&start_sha=beca27ae7009819e26d572b307aa909bb5805095)

**Comment by NicolasGensollen** [comment]
changed this line in [version 14 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605619331&start_sha=1d0dbba2d5bc0f7be5f201421d4f42cbb821adbb#15761437d39369b398641b30e6eba7ec0289a59c_144_145)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>1c36cd3b - check model compat before initialization</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605619331&start_sha=1d0dbba2d5bc0f7be5f201421d4f42cbb821adbb)

**Comment by NicolasGensollen** [comment]
Done in 1c36cd3b

**Comment by NicolasGensollen** [comment]
@emaheux I think I addressed all your points.

A few comments about the commits I made this morning:

In c38e5099 I moved `validate_compatibility_of_dataset` in base classes (`AbstractModel` and `GenericModel`) since this method seems to apply to all Leaspy models. I also made some changes to the `dimension` and `features` attributes which are a little difficult to work with because they are strongly coupled, but can be `None` if the model hasn't been initialized. I proposed to make them both private and added getters and setters which have the responsibility of maintaining the coherence between them. Let me know whether this sounds reasonable or not from your perspective.

In 1d0dbba2 I went one step further (and probably too far compared to the scope I initially wanted to give to this PR...). Since both the `AbstractModel` and the `GenericModel` now share some common logic, I created a new class `BaseModel` (name can change of course...) from which both `AbstractModel` and `GenericModel` now inherit. For the moment, this class doesn't hold much, but it does unify both model inheritance trees and could serve as a starting point for the refactoring we discussed this week.

Let me know what you think. We can either include this in this PR, have it as a separate follow-up PR, or remove it altogether.

**Comment by NicolasGensollen** [comment]
Adressed in c38e5099 and 1d0dbba2

**Comment by NicolasGensollen** [comment]
Adressed in c38e5099 and 1d0dbba2

**Comment by NicolasGensollen** [comment]
resolved all threads

**Comment by NicolasGensollen** [comment]
marked this merge request as **ready**

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>56ccf268 - Add docstring to BaseModel</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605655098&start_sha=1c36cd3bdac05a0eeb8860aaee72e4f250cf19c7)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>66eae4b3 - Add docstring to BaseModel</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605762904&start_sha=56ccf268fd0e9240f5f74e982a584cd090a38beb)

**Comment by NicolasGensollen** [comment]
added 1 commit

<ul><li>589de967 - Add docstring to BaseModel</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/90/diffs?diff_id=605808078&start_sha=66eae4b3f0eead37cde31a674109caa486c752bc)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Mar 2, 2023, 16:56

Nice refact! Seems perfect to me :slight_smile: 
I let you merge the squashed version!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Mar 2, 2023, 16:56

approved this merge request

**Comment by NicolasGensollen** [comment]
Alright ! Thanks for the review @emaheux ! 
Merging then...

**Comment by NicolasGensollen** [comment]
mentioned in commit 1d634174d3f4749b1e449ab66744e49b054ead7b
